File: Patrones Estructurales/Composite/composite.js
// 1. Component (La interfaz común)
class Componente {
    obtenerPrecio() {
        throw new Error("Método no implementado");
    }
}

// 2. Leaf (Hoja - El objeto simple)
class Producto extends Componente {
    constructor(nombre, precio) {
        super();
        this.nombre = nombre;
        this.precio = precio;
    }

    obtenerPrecio() {
        return this.precio;
    }
}

// 3. Composite (Contenedor - Contiene hojas u otros contenedores)
class Caja extends Componente {
    constructor() {
        super();
        this.hijos = []; // Lista de componentes
    }

    agregar(componente) {
        this.hijos.push(componente);
    }

    obtenerPrecio() {
        let total = 0;
        // Recorre sus hijos y suma sus precios (recursividad)
        for (const hijo of this.hijos) {
            total += hijo.obtenerPrecio();
        }
        return total;
    }
}

// Uso
const telefono = new Producto("iPhone", 1000);
const audifonos = new Producto("AirPods", 200);

const cajaPequeña = new Caja();
cajaPequeña.agregar(audifonos);

const cajaGrande = new Caja();
cajaGrande.agregar(telefono);
cajaGrande.agregar(cajaPequeña); // Una caja dentro de otra

console.log(`Precio total: $${cajaGrande.obtenerPrecio()}`); 
// Salida: Precio total: $1200

File: Patrones Estructurales/Composite/composite.py
from abc import ABC, abstractmethod

# 1. Component
class Componente(ABC):
    @abstractmethod
    def obtener_precio(self) -> float:
        pass

# 2. Leaf
class Producto(Componente):
    def __init__(self, nombre, precio):
        self.nombre = nombre
        self.precio = precio

    def obtener_precio(self) -> float:
        return self.precio

# 3. Composite
class Caja(Componente):
    def __init__(self):
        self._hijos = []

    def agregar(self, componente: Componente):
        self._hijos.append(componente)

    def obtener_precio(self) -> float:
        total = 0
        for hijo in self._hijos:
            total += hijo.obtener_precio()
        return total

# Uso
telefono = Producto("iPhone", 1000)
cargador = Producto("Cargador", 50)

caja_principal = Caja()
caja_principal.agregar(telefono)
caja_principal.agregar(cargador)

print(f"Precio total: ${caja_principal.obtener_precio()}")